### ISSUE TITLE
about the computation of variance

### ISSUE THREAD

**Issue by ShellingFord221** [pmorerio/dl-uncertainty#2]
Hi, these days I find an old paper _Estimating the Mean and Variance of the Target Probability Distribution_ to support the idea about considering mean and variance as two independent output of the network. I think that Eq. 11-14 may be the reason why you regress the images (though I don't know why).
<img width="397" alt="image" src="https://user-images.githubusercontent.com/40685673/70622090-19923980-1c56-11ea-823f-0c2fd0d9933a.png">
Besides, I'm still wondering that what is the difference between taking the variance as a direct output and computing the variance based on the sampled results. Do they actually have the same meaning (or same value)? Thanks!

**Comment by pmorerio** [comment]
> you regress the images (though I don't know why).

Simply because I am interested in a regression problem. In the NIPS paper you can also find the case of classification.

**Comment by pmorerio** [comment]
> what is the difference between taking the variance as a direct output and computing the variance based on the sampled results. Do they actually have the same meaning (or same value)?

According to my understanding of the NIPS paper, they do not have the same meaning, the former being aleatoric uncertainty and the latter being epistemic (I have implemented both separately and combined).

**Comment by ShellingFord221** [comment]
Sorry that maybe I misuse the word. I mean that Eq. 11-14 may be the reason why you use an encoder-decoder framework.

**Comment by ShellingFord221** [comment]
Besides, in combined/model.py line 103~109, it seems only aleatoric uncertainty takes part in the loss function. So why is this a combination of both two uncertainties? Does 'Combining Aleatoric and Epistemic Uncertainty in One Model' mean that we can get two kinds of uncertainties of the model, rather than considering both of them in the loss function to see how do these two uncertainties affect the model? (Although the predictive uncertainty for pixel y is the combination of aleatoric and epistemic, i.e. Eq. 9)

**Comment by pmorerio** [comment]
> So why is this a combination of both two uncertainties?

Only aleatoric uncertainty takes part in the loss function, since epistemic uncertainty estimation is simply done by turning on dropout at test time. I sample 20 nets, and compute mean and variance of their predictions. In my understanding this is the way of combining the two uncertainties.

**Comment by ShellingFord221** [comment]
Hi, I have read another paper _Bayesian SegNet: Model Uncertainty in Deep Convolutional Encoder-Decoder Architectures for Scene Understanding_, in which there is a Bayesian SegNet with an encoder-decoder architecture. Encoder encodes features and decoder upsamples them with dropout layer inserted. So I wonder that is your encoder-decoder architecture just for the specific network (SegNet for example), or is it designed to get mean and variance of the output? Thanks!

**Comment by pmorerio** [comment]
> So I wonder that is your encoder-decoder architecture just for the specific network (SegNet for example), or is it designed to get mean and variance of the output?

Hi, sorry but again I am not completely sure about the meaning of your question. I also design an encoder-decoder for regression purposes. The decoder has two heads to regress mean as well as variance in the case of aleatoric uncertainty.

**Comment by ShellingFord221** [comment]
I'm sorry that maybe I don't make my purpose clear. Put simply, how can I implement predictive mean and predictive variance in a classification task without encoder-decoder architecture?

**Comment by pmorerio** [comment]
Oh ok, got it.

I think you should first define whether your are interested in 1) Heteroscedastic (Aleatoric) Uncertainty or 2) Epistemic Uncertainty.

1.  You should check section 3.3 of the paper. I have not read carefully that part.
2. I think you can obtain it by performing dropout at test time and averaging as suggested by equation (3).

In any case i suggest to study carefully the whole paper. Consider contacting the authors as well, they can help you better I guess.

Best,
P.

**Comment by ShellingFord221** [comment]
Thank you for your suggestion. For your suggestion 2, the epistemic uncertainty can be easily captured by performing dropout and averaging the outputs, but modeling aleatoric uncertainty in classification also needs to output predictive mean as well as predictive variance just like regression task (please see Eq. 6), so I'm interested in your method about using an encoder-decoder architecture to do this, and also wondering that whether I can get mean and variance without it. I think in this step, both classification task and regression task are nealy the same. After getting mean and variance, we can use it to disturb logit vector to get the form of loss attenuation of a classification scenario, i.e. Eq. 12 (I'm pretty clear about this part). 
So in a word, I just want to know how to 'use a single network to transform the input x, with its head split to predict both y^ as well as σ^2'? Thanks!

**Comment by pmorerio** [comment]
> I think in this step, both classification task and regression task are nealy the same

Not so sure about this.

> So in a word, I just want to know how to use a single network to transform the input x, with its head split to predict both y^ as well as σ^2'?

For _regression_ I have implemented it in `model.py` in the aleatoric-uncertainty folder. I guess there are other ways of doing that. For _classification_, honestly I don't know.
